uniffi_core/src/ffi/rustfuture/mod.rs

```diff
@@ -30,6 +30,32 @@ pub enum RustFuturePoll {
 /// to continue progress on the future.
 pub type RustFutureContinuationCallback = extern "C" fn(callback_data: u64, RustFuturePoll);
 
+/// This marker trait allows us to put different bounds on the `Future`s we support, based on
+/// `#[cfg()]` configuration.
+pub trait UniffiCompatibleFuture<T, E>: Future<Output = Result<T, E>> {}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl<T, F, E> UniffiCompatibleFuture<T, E> for F where F: Future<Output = Result<T, E>> + Send {}
+
+/// # Safety:
+///
+/// WASM32 is a single threaded environment. However, in a browser there do exist
+/// [`WebWorker`](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers)s which do
+/// not share memory or event-loop with the main browser context.
+///
+/// The most common source of asynchrony,
+/// [`JsPromise`](https://github.com/rustwasm/wasm-bindgen/blob/main/crates/futures/src/lib.rs)
+/// is not `Send` because:
+///
+/// 1. `T`` and `E`` are both `JsValue`
+/// 2. `JsValue` may contain `JsFunction`s, either as a function themselves or an object containing functions.
+/// 3. Functions cannot be serialized and sent to `WebWorker`s.
+///
+/// The polling and waking happens in the binding language, i.e. the main browser context, i.e. the same thread
```

Comment:
"polling and waking happens in the binding language, i.e. the main browser context" - I don't know much about web workers, what prevents users from calling UniFFI stuff from within one?

Reply:
Good question, can you poll a future in a web worker?  I think that's still safe, since the model seems to be that you send messages across threads rather than share the futures.  Maybe it's more accurate to say "polling and waking happens in the same thread the future was created in".

Reply:
I have enumerated the various combinations, which are either possible, or impossible for different reasons.